articles/build-rules.md

```diff
@@ -0,0 +1,89 @@
+---
+authorName: Vassilis Rizopoulos
+authorGithubUsername: damphyr
+---
+# TBD
+
+This article discusses the application of the [Separation of Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) concept in Build & CI systems. 
+
+It is layed out as a set of rules to follow when creating build and CI infrastructure for your project. 
+
+Think of each rule as a [T.A.R.D.I.S.](https://en.wikipedia.org/wiki/TARDIS): they are bigger on the inside. 
+
+## Distinctions, distinctions
+
+To identify the separate concerns we should provide definitions for the terms: 
+
+A **build system** performs transformations in sequence according to a predetermined dependency chain to create artifacts. A subset of this is the compilation of sources to binaries.
+
+A **build management system** coordinates build system(s).
+
+### Specificity
+
+A build system is highly project specific. It is affected by toolchains, project conventions and generally can only run on a specially configured host (i.e. development environment).
+
+A build management system can run everywhere as long as it can start the build system on the appropriate host. The specificity of a build management system is limited to the number of version control systems it supports. Although theoretically it doesn't have to provide version control support it is a given that such support will be provided in the minimum feature set.
+
+### The Makefile is the build system
+
+Make, rake, Ant, MSBuild, graddle, grunt etc. **are not build systems**<sup>1</sup>. You create build systems with them. 
+
+Jenkins, TeamCity, TravisCI, BuildBot, TFS etc. are build management systems.
+
+## The Rules
+
+### 1. I am the build server
+
+Rule #1 requires that the build server follows the exact same steps as any other developer. 
+
+Expressed the other way around: Every developer has to be able to recreate locally the complete build process without deviations given the development environment and the correct version of the source tree.
+
+### 2. When the build server says no, it means no!
+
+Rule #2 says that if a build server marks a build as broken, then the build is broken. Drop everything and read the logs. 
+
+There is no "it works for me", your build server is Judge Dredd: judge, juror and executioner.
+
+You can only adhere to this rule if you followed Rule #1
+
+### 3. IDEs are the enemy AKA F5 is not a build process
+
+Means that if you drive your development process from an IDE there is no way you can adhere to the Build Rules. 
+
+This rule has major complications on the development environment, ties directly into the subject of allowing your developers to use whatever tools they feel comfortable with and adherence to it is actively sabotaged by Visual Studio.
```

Comment:
This rule has major complications on the development environment. It ties directly into the subject of allowing your developers to use whatever tools they feel comfortable with. While this freedom of choice is standard for Java developers, it has not reached the .NET world yet where Visual Studio is hard coupled to MSBuild.

Reply:
No, no, "actively sabotaged" is exactly what I meant :)

But it is no longer true (the source material was written in 2014) - VS since the 2015 version allows us to even integrate third party build systems in the IDE.